### ISSUE TITLE
Health check

### ISSUE THREAD

**Issue by bhugot** [Azure/azure-functions-durable-extension#2223]
Hello,
It's a feature that would be cool is to have a way to have a healthcheck for durable task. Currently we have an implementation using ListInstancesAsync but it's really not efficient with azure storage at least, as we need to use continuation token to get first potential Failed Orchestration. Maybe there is a better way to do it?

Regards

**Comment by cgillum** [comment]
When using ListInstancesAsync, have you tried using the runtime status filter?

As an alternative to ListInstancesAsync, you can also use [Event Grid Publishing](https://docs.microsoft.com/azure/azure-functions/durable/durable-functions-event-publishing) to receive notifications about failed instances proactively. Would this be helpful?

**Comment by bhugot** [comment]
Yes i m filtering with status and date too lower the total number of item.

**Comment by cgillum** [comment]
What about the  [Event Grid Publishing](https://docs.microsoft.com/azure/azure-functions/durable/durable-functions-event-publishing) suggestion I mentioned?

Alternatively, would it make sense to use [App Insights queries](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-diagnostics#application-insights) for health monitoring? This option would be far more efficient compared to querying Azure Storage via our `ListInstancesAsync` API.

**Comment by bhugot** [comment]
Maybe event grid could it but I m not sure. What WE need is the current state so how many orchestration are in failed state. So it mean to have a state of each of them. But I will check each of your solutions.

**Comment by cgillum** [comment]
I see. We don't have a built-in mechanism for automatically summarizing the current state of all orchestration instances. The best option we have for this is using `ListInstancesAsync`. We could consider adding a new feature for this, but we'd need to design it carefully and understand the use-cases. Would you be willing to describe in more detail how you want to use such a feature?

/cc @lilyjma

**Comment by mpaul31** [comment]
I would suggest using event grid like @cgillum suggested and applying a filter for only publishing failed orchestrations. This is our current strategy and works great.

**Comment by olitomlinson** [comment]
App Insights is a great way to know _when_ an orchestration has gone into a failed state, but its not got good at giving you an aggregate count of how many instances are in a failed state _right now_ .

If you really need the count, a straight forward way to do this is as follows :

- Use the Event Grid lifecycle hooks (restricted to `failed` state for efficiency)
- Receive the event hook in a Function App (ideally a different function App to the one where your Durable Functions are running)
- Write the event into an Azure Table (or similar tech, whatever is appropriate for your use-case)
- In the Function App, write a Cron triggered function, that retrieves the total count of the number of rows that are present in that Table.
- Publish that count as a metric to App Insights. This is your count of _current_ failed instances.
- Create an Azure Monitor Alert if the count ever > 0
- When you have triaged those particular failures, go into the Azure Table and clear them out so it is in sync with the actual system. (not glamorous I know, but this could be automated with a bit of creativity)

I used to do exactly this, so I know it can work.
